src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Emitter/CoreBindingHelpers.cs

```diff
@@ -901,6 +901,15 @@ private bool EmitBindImplForMember(
                         }
                     case ComplexTypeSpec complexType:
                         {
+                            // Early detection of types we cannot bind to and skip it.
+                            if (!_typeIndex.HasBindableMembers(complexType) &&
+                                !_typeIndex.GetEffectiveTypeSpec(complexType).IsValueType &&
+                                complexType is not CollectionSpec &&
+                                ((ObjectSpec)complexType).InstantiationStrategy == ObjectInstantiationStrategy.ParameterizedConstructor)
```

Comment:
Can you explain all these conditions?  I only understand the first, why the others?

Reply:
`complexType is not CollectionSpec` we have special handling to the collections so we should not change how we do it especially it should work in most cases. look at https://github.com/dotnet/runtime/blob/e4f93dbee0bc3f17f5ac2fad34e75f38865e866d/src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Emitter/CoreBindingHelpers.cs#L1146 for more details.

`((ObjectSpec)complexType).InstantiationStrategy == ObjectInstantiationStrategy.ParameterizedConstructor)` is needed because if it is different ObjectInstantiationStrategy value means we can still create and handle the object. look at https://github.com/dotnet/runtime/blob/e4f93dbee0bc3f17f5ac2fad34e75f38865e866d/src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Emitter/CoreBindingHelpers.cs#L1170 to see that.

Reply:
`_typeIndex.GetEffectiveTypeSpec(complexType).IsValueType` can be handled too, https://github.com/dotnet/runtime/blob/e4f93dbee0bc3f17f5ac2fad34e75f38865e866d/src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Emitter/CoreBindingHelpers.cs#L933 shows what we do with that.

Reply:
I was trying to scope the change to exact failure cases to reduce any risk of breaking anything.

Reply:
That is because we want use `_typeIndex` which is created and filled in the emitter. If we want to do the detection earlier, we'll need to do more involving refactoring which will make the change risky for 9.0. Note, the parser fills some data structures which expect which expect to see the type specs supported by `_typeIndex` later during the emission. so refactoring will not be simple IMO.

Reply:
I see - given that it seems like the parser isn't even capable of completely deciding which types will be handled.  It can only decide which *may* be handled and we filter further during emit.  That feels like something we'd need to fix to address other issues.